lib/kube/proxy/forwarder.go

```diff
@@ -652,16 +664,66 @@ func (f *Forwarder) setupForwardingHeaders(ctx *authContext, sess *clusterSessio
 	req.Header.Add("X-Forwarded-Host", req.Host)
 	req.Header.Add("X-Forwarded-Path", req.URL.Path)
 
+	return nil
+}
+
+// setupImpersonationHeaders sets up Impersonate-User and Impersonate-Group headers
+func setupImpersonationHeaders(log log.FieldLogger, ctx *authContext, headers http.Header, bearerToken string) error {
+	var impersonateUser string
+	var impersonateGroups []string
+	for header, values := range headers {
+		if !strings.HasPrefix(header, "Impersonate-") {
+			continue
+		}
+		switch header {
+		case ImpersonateUserHeader:
+			if impersonateUser != "" {
+				return trace.AccessDenied("%v, user already specified to %q", ImpersonationRequestDeniedMessage, impersonateUser)
+			}
+			if len(values) == 0 || len(values) > 1 {
+				return trace.AccessDenied("%v, invalid user header %q", ImpersonationRequestDeniedMessage, values)
+			}
+			impersonateUser = values[0]
+			if _, ok := ctx.kubeUsers[impersonateUser]; !ok {
+				return trace.AccessDenied("%v, user header %q is not allowed in roles", ImpersonationRequestDeniedMessage, impersonateUser)
+			}
+		case ImpersonateGroupHeader:
+			for _, group := range values {
+				if _, ok := ctx.kubeGroups[group]; !ok {
+					return trace.AccessDenied("%v, group header %q value is not allowed in roles", ImpersonationRequestDeniedMessage, group)
+				}
+				impersonateGroups = append(impersonateGroups, group)
+			}
+		default:
+			return trace.AccessDenied("%v, unsupported impersonation header %q", ImpersonationRequestDeniedMessage, header)
+		}
+	}
+
+	impersonateGroups = utils.Deduplicate(impersonateGroups)
+	if impersonateUser == "" {
+		if len(ctx.kubeUsers) == 1 {
+			for user := range ctx.kubeUsers {
+				impersonateUser = user
+				log.Debugf("Setting impersonate user to %v: defaulting single value from roles", user)
+				break
+			}
+		} else {
```

Comment:
Why is holding a single user given special treatment?  This seems like it could create confusing behavior.

E.G. `alice` has role `dev`.  Alice impersonates herself by default.  Bob adds `admin` user to the role `dev`.  Now alice impersonates `admin`.  Bob then adds `testing` user to `dev`.  Alice is now back to impersonating herself.  From alice's point of view, this is very confusing.

Reply:
This is to allow a default: it does not matter how many roles Alice holds, as long as the resulting role set only specifies one user to impersonate, the system will default to it. In your example, if `Bob` adds `admin` user to the role `dev`, the system will not be able to choose a default, because there are two users in the set, and will ask `alice` to pick one.

Reply:
> the system will not be able to choose a default, because there are two users in the set, and will ask Alice to pick one.

Seems that `impersonateUser = ctx.User.GetName()` always runs unless `len(ctx.kubeUsers) == 1`... maybe I'm misunderstanding, but that looks to me like alice will initially impersonate herself, start impersonating the first user added to any role she holds, then go back to impersonating herself as soon as any role adds an additional user.

Reply:
> the system will not be able to choose a default, because there are two users in the set, and will ask Alice to pick one.

Seems that `impersonateUser = ctx.User.GetName()` always runs unless `len(ctx.kubeUsers) == 1`... maybe I'm misunderstanding, but that looks to me like alice will initially impersonate herself, start impersonating the first user added to any role she holds, then go back to impersonating herself as soon as any role adds an additional user.

Reply:
yes, you are right, that would be the effect, I'm wondering what would be a better default then:

* Alice would not be impersonating any user by default unless specified explicitly
* Alice would be impersonating herself by default at all times, unless specified otherwise and impersonating herself explicitly will be allowed
* Alice would be impersonating only users set in the `kubernetes_users` from now on (if there is only one user specified) and no other impersonation will be allowed